cpp/src/strings/replace/backref_re.cu

```diff
@@ -0,0 +1,231 @@
+#include <cudf/strings/string_view.cuh>
+#include <cudf/strings/char_types/char_types.hpp>
+#include <cudf/strings/replace_re.hpp>
+#include <strings/utilities.hpp>
+#include <strings/utilities.cuh>
+#include <strings/regex/regex.cuh>
+
+#include <regex>
+
+namespace cudf
+{
+namespace strings
+{
+namespace detail
+{
+namespace
+{
+
+/**
+ * @brief Parse the back-ref index and position values from a given replace format.
+ *
+ * The backref numbers are expected to be 1-based.
+ *
+ * Returns a modified string without back-ref indicators.
+ * ```
+ * Example:
+ *    for input string:    'hello \2 and \1'
+ *    the returned pairs:  (2,6),(1,11)
+ *    returned string is:  'hello  and '
+ * ```
+ */
+std::string parse_backrefs( std::string const& repl, std::vector<thrust::pair<size_type,size_type> >& backrefs )
+{
+    std::string str = repl; // make a modifiable copy
+    std::smatch m;
+    std::regex ex("(\\\\\\d+)"); // this searches for backslash-number(s); example "\1"
+    std::string rtn;             // result without refs
+    size_type byte_offset = 0;
+    while( std::regex_search( str, m, ex ) )
+    {
+        if( m.size()==0 )
+            break;
+        std::pair<size_type,size_type> item;
+        std::string bref = m[0];
+        size_type position = static_cast<size_type>(m.position(0));
+        size_type length = static_cast<size_type>(bref.length());
+        byte_offset += position;
+        item.first = std::atoi(bref.c_str()+1); // back-ref index number
+        CUDF_EXPECTS( item.first > 0, "Back-reference numbers must be greater than 0");
+        item.second = byte_offset;              // position within the string
+        rtn += str.substr(0,position);
+        str = str.substr(position + length);
+        backrefs.push_back(item);
+    }
+    if( !str.empty() ) // add the remainder
+        rtn += str;    // of the string
+    return rtn;
+}
+
+
+/**
+ * @brief This functor handles replacing strings by applying the compiled regex pattern
+ * and inserting the at the backref position indicated in the replacement template.
+ *
+ * The logic includes computing the size of each string and also writing the output.
+ *
+ * The stack is used to keep progress on evaluating the regex instructions on each string.
+ * So the size of the stack is in proportion to the number of instructions in the given regex pattern.
+ *
+ * There are three call types based on the number of regex instructions in the given pattern.
+ * Small to medium instruction lengths can use the stack effectively though smaller executes faster.
+ * Longer patterns require global memory. Shorter patterns are common in data cleaning.
+ *
+ */
+template<size_t stack_size>
+struct backrefs_fn
+{
+    column_device_view const d_strings;
+    reprog_device prog;
+    string_view const d_repl;   // string replacement template
+    thrust::pair<size_type,size_type>* d_backrefs; // ref/pos pairs
+    size_type backref_count;
```

Comment:
We cannot use `std::vector` or even a `device_vector` inside device code but we can pass and use iterators instead of using the ptr and count.
I've updated the code to use `for_each` with iterators.